compiler/rustc_errors/src/emitter.rs

```diff
@@ -398,13 +395,24 @@ pub trait Emitter {
                 // and it needs an "in this macro invocation" label to match that.
                 let redundant_span = trace.call_site.contains(sp);
 
-                if !redundant_span && matches!(trace.kind, ExpnKind::Macro(MacroKind::Bang, _))
-                    || always_backtrace
-                {
+                if !redundant_span || always_backtrace {
                     new_labels.push((
                         trace.call_site,
                         format!(
-                            "in this macro invocation{}",
+                            "in {}{}",
+                            match trace.kind {
+                                ExpnKind::Macro(MacroKind::Attr, _) =>
+                                    "this procedural macro expansion".to_string(),
+                                ExpnKind::Macro(MacroKind::Derive, _) =>
+                                    "this derive macro expansion".to_string(),
+                                ExpnKind::Macro(MacroKind::Bang, _) =>
+                                    "this macro invocation".to_string(),
+                                ExpnKind::Inlined => "the inlined copy of this code".to_string(),
+                                ExpnKind::Root => "in the crate root".to_string(),
+                                ExpnKind::AstPass(kind) => kind.descr().to_string(),
+                                ExpnKind::Desugaring(kind) =>
+                                    format!("this {} desugaring", kind.descr()),
+                            },
```

Comment:
How about using a cow for this? It's already imported.
```suggestion
                            {
                              let place: Cow<'static, str> = match trace.kind {
                                  ExpnKind::Macro(MacroKind::Attr, _) => "this procedural macro expansion".into(),
                                  ExpnKind::Macro(MacroKind::Derive, _) => "this derive macro expansion".into(),
                                  ExpnKind::Macro(MacroKind::Bang, _) => "this macro invocation".into(),
                                  ExpnKind::Inlined => "the inlined copy of this code".into(),
                                  ExpnKind::Root => "in the crate root".into(),
                                  ExpnKind::AstPass(kind) => kind.descr().into(),
                                  ExpnKind::Desugaring(kind) => format!("this {} desugaring", kind.descr()).into(),
                              };
                              
                              place
                            },
```